### ISSUE TITLE
Exploring Ethereum on libp2p

### ISSUE THREAD

**Issue by bigs** [libp2p/libp2p#33]
# Ethereum projects on libp2p

Starting a thread so we can keep track of questions regarding what it might take to get Ethereum (present and future) operating over libp2p.

## General Interests

Ethereum 1's codebase already features a lot of libp2p's functionality, but they'd like to remove some of that burden and expand the capabilities. Furthermore, with Ethereum 2 research beginning as a relatively green-field project, they would like to identify the most stable/performant pieces of libp2p to maximize ease of development vs future pain.

- Swappable transports
- Zero round-trip handshakes (a la DTLS/QUIC)
- Stronger encrypted transports
- Attack mitigation (DDOS, timing attacks, etc. Obviously a big one!)
  - Security audits planned/executed?
  - Node identity and authentication (perhaps verifiable from sources outside of peer network)
  - Peer selection strategies (is random-walk appropriate)
  - In degree vs out degree measurement for connection rebalancing
- Connection management (likely relevant to ongoing peer swarm refactoring)
  - Pluggable `ConnManager`?
- Gossipsub
  - Performance characteristics
  - What are its IPFS/PL use cases? Will this implementation stick around and get the love it needs to become a core part of either Ethereum or PL infrastructure.
- Potential IPFS/Ethereum interactions
- Python/other language support
  - libp2p daemon is likely highest priority

## Sharding-specific Interests

- DHT persistence, speed of bootstrapping
  - When rejoining the network, are your peers good/connected?
  - DHT routing table persistence not yet landed
- How might we map shards to libp2p?
  - Separate peer network per shard
  - Single channel that dispatches clients to separate peer networks
  - Multiple channels overlayed over a single shared peer network
- How can we optimize balance of flood/gossipsub propagation vs active connections?
- Support for authentication in flood/gossipsub?
- Fast shard jumping
  - Can we (perhaps at a slower interval) keep a small cache of DHT routing table for other shards?
  - Can we keep an open connection to some small number of peers in shards we're not actively participating in to speed this switch
  - QUIC transport clearly a huge boon here

I'll edit this initial post as I get more clarity on requirements and requests!

Attn: @mgoelzer, @JustinDrake

**Comment by ghost** [comment]
cc @fjl

**Comment by jamesray1** [comment]
I'm still reading and learning, but generally, it would be good to know the pros and cons of using rust-libp2p as is for sharding, vs. using gossipsub (instead of floodsub) and other enhancements.

**Comment by ghost** [comment]
@jamesray1 On rust-libp2p, this is something the Parity folks are developing.  I believe it does not implement the full [libp2p specs](https://github.com/libp2p/specs), but it is extremely efficient on low-memory systems like SoCs.  @tomaka can you explain in more detail.  Based on what I know about the sharding project, I think the Go implementation will serve you better.  As you point out, the gossipsub stuff is only in Go at present; the Rust implementation would be floodsub.

With regard to the pros and cons of the two pubsub implementations, let me punt that to @Stebalien or @whyrusleeping.  I think they will give you a better answer than I can.

**Comment by jamesray1** [comment]
@mgoelzer thanks for your recommendation, I will look into gossipsub more closely. In that case rather than porting stuff from the Go implementation, which can be unsafe, I will probably develop a gossipsub implementation in Rust. I have talked with Pierre about this at https://github.com/libp2p/rust-libp2p/issues/187.

**Comment by ghost** [comment]
@jamesray1 Great!  Note that some of the folks I mentioned are on vacation for next ~7 days, so you might not get replies quickly.  We won't forget though.

**Comment by mhchia** [comment]
@bigs @mgoelzer Thank you for the conclusion! It's pretty exact and clear.

I'm Kevin from Ethereum and I also attended the meeting that day. I would like to add some thoughts and updates we currently have in mind. (might change, the concept is still under refining).

Currently, we are working on doing poc on [gossipsub](https://github.com/libp2p/go-floodsub/pull/67/). One possible scheme we are trying is, messages are mostly propagated through the pubsub, including transactions, proposals, shard blocks(collations), and "listeningShards" of nodes.
In detail, there is one topic "listeningShards" for nodes to publish the shards they are currently listening to. So every node should subscribe to this topic, maintain the "listeningShards" table mapping from all nodes' peerID to their "listeningShards", and publish its currently "listeningShards" every time the node listens to a shard. Besides, every shard corresponds to a topic, like shard 0 to "shard_0", shard 1 to "shard_1", ...etc.
If a node is to listen to a shard, what it needs to do is 
1. Connect with the peers which subscribe to the shards it is to connect to in the "listeningShards" table.
2. Subscribe to the topic corresponding to that shard, say, "shard_k".

In this scheme, for "Fast shard jumping",  we might also need
- the latency to fetch connection information of nodes to be low, which might be possibly improved by doing `dht.findPeer(peerID)` in advance, "caching in routing table", "QIUC" as you mentioned.
- some design in maintaining peers. For example, if every node connects mostly with the nodes which listen to the same shard, it seems the network is easier to be partitioned. Though it should be mitigated by having bootstrap nodes in every shard.

Please feel free to correct me or give any feedback:) I'm really grateful for the help from you guys!

**Comment by jamesray1** [comment]
@mhchia 

> messages are mostly propagated through the pubsub, including transactions, proposals, shard blocks(collations), and "listeningShards" of nodes.

This adds a lot of bandwidth overhead by flooding peers, using gossipsub would be more efficient.

> some design in maintaining peers. For example, if every node connects mostly with the nodes which listen to the same shard, it seems the network is easier to be partitioned. Though it should be mitigated by having bootstrap nodes in every shard.

Yes, we need to be careful on becoming reliant on a defacto set of peers for each shard. Having bootstrap nodes also seems concerning, as they would be a point of centralization and thus be easier to attack. And without gossipsub then it also seems like it would be easier to depend on a bootstrap node or a few centralized nodes for messages.

Speaking more generally than messaging, it seems that other IPFS modules that rust-libp2p doesn't have are useful, e.g. for future-proofing and flexibility, multihash would be useful for flexibility instead of just using sha3; and generally the other modules are handy for p2p. Since Ethereum can be used for any economic or governance activity, we should be flexible for network design. For performance, perhaps the rust-libp2p implementation can continue with their approach, but forking that and developing more IPFS features seems useful, although it may be a big project (again, I'm still reading).

**Comment by dignifiedquire** [comment]
Just a note regarding multihash, there are rust implementations for all the major formats that IPFS is using, in case you missed them. 

- [rust-multiaddr](https://github.com/multiformats/rust-multiaddr)
- [rust-multihash](https://github.com/multiformats/rust-multihash)
- [rust-mutlibase](https://github.com/multiformats/rust-multibase)
- [rust-cid](https://github.com/ipld/rust-cid)

**Comment by whyrusleeping** [comment]
@mhchia is your work up on github anywhere? I'd love to peek at it and see how things look :)

On the rust front, I just watched a demo of polkadot running and syncing over rust-libp2p. @tomaka is doing some great work there

If there are any specific examples, bits of documentation or anything that would make life easier for anyone here, please let me know.

**Comment by mhchia** [comment]
@whyrusleeping Yes, the [poc code](https://github.com/mhchia/go-libp2p/tree/poc-testing/examples/minimal) forked `go-libp2p` and started from the examples, so there are a lot of references to the code in existing examples.(logics related to topic subscription is [here](https://github.com/mhchia/go-libp2p/blob/poc-testing/examples/minimal/shardmanager.go)) It is pretty simple now and please kindly tolerate my spaghetti since I'm not so familiar with golang 😂 I will really appreciate if there is any feedback.

Previously I did tests on sending 1MB shard blocks from nodes to nodes, and the latency for nodes to `dht.findPeer`, trying to figure out the capability of `floodsub`. It seems to work great, I will do it further and investigate more on our needs, to see if both matches. I actually have some questions, is there a specification for the limit of current floodsub/gossipsub? Just wonder how is the limit of message size set?(in floodsub it seems 1MB) also the size of queue for RPC?

 What I'm currently doing is make the code more fittable with `gossipsub`, including tracing 
 more into and understanding more about `gossipsub`, since it is previously tested only with `floodsub`, and also making the code more complete with CLI, ...etc. To conclude, I still need to be more familiar with `gossipsub`.

Last but not least, though it seems the code,  [RFC and papers](https://github.com/vyzo/research-pubsub/blob/a6cfdd3fec9d595a3f4cb56e94642e77b4dca13f/epidemic-pubsub.md#joining-the-overlay) and the [document](https://github.com/vyzo/gerbil-simsub) are available, I think it will be helpful if there is any more resource. Will see if there is anything needed after I finished going through these existing resources:)

Big thanks for staying tune with this thread

**Comment by jamesray1** [comment]
@dignifiedquire, thanks, yeah I've seen the first two, but not the last two. Looks like it will just be a matter of developing gossipsub, and integrating other modules as needed from rust-libp2p and the others you mentioned.

**Comment by whyrusleeping** [comment]
@mhchia 

>  I actually have some questions, is there a specification for the limit of current floodsub/gossipsub?

It's actually not specified in docs for floodsub, but we use 1MB all over different libp2p protocols as a bit of a DoS prevention mechanism (don't want peers spamming our memory by sending us tons of large objects, forcing us to allocate buffers for them to receive the message).

The value is set right here: https://github.com/libp2p/go-floodsub/blob/master/comm.go#L29 (and i'm a bit annoyed that its not pulled out into a nice constant).

If you need it to be larger, depending on how large it is, we can just make it bigger (if, say, 2MB or 4MB would fix any issue), or we can make it configurable. Changing this number does run into issues with other clients who do not have the same number as you potentially dropping your messages if they are bigger than they support, so interop with existing peers might be weird if you increase that (this is likely not an issue).

I'll see if I can find time to look at your code and send you some pointers (or i'll see if anyone on our team is interested in doing some code review for you)

**Comment by mhchia** [comment]
@whyrusleeping 
> It's actually not specified in docs for floodsub, but we use 1MB all over different libp2p protocols as a bit of a DoS prevention mechanism (don't want peers spamming our memory by sending us tons of large objects, forcing us to allocate buffers for them to receive the message).

Sorry for the late reply and Thank you a lot for the instruction! I think `1 MB` should be a big enough number. And please only look at the code if you have spare time. That is not in a hurry.

I also wanted to ask how did you guys measure/determine the size of the channel for RPC, like based on some simulation or calculation?  For example, I tried to keep publishing many `1 MB` messages from one node really fast. Some warning messages showed up and some of the messages were not sent successfully. 
```bash
pubsub: dropping message to peer <peer.ID S5Qmci>: queue full pubsub.go:479
pubsub: Can't deliver message to subscription for topic shardCollations_0; subscriber too slow pubsub.go:478
```
I changed the size of [channel for handling sending messages](https://github.com/libp2p/go-floodsub/blob/master/pubsub.go#L186), and the size of [channel for each topic](https://github.com/libp2p/go-floodsub/blob/master/pubsub.go#L297), to workaround the two issues respectively before. I’m wondering if these sizes are possibly configurable. Or instead, is it just my bandwidth problem, and I should not modify these settings and should slow the publishing rate?

Really appreciate that you spend time on this :)

**Comment by whyrusleeping** [comment]
@mhchia hrm... those numbers were (admittedly somewhat arbitrarily) selected as being reasonable buffers for short bursts of message sending that is higher than outgoing bandwidth. If your application runs into that consistently, then it is likely to run into that at some point even if you bump the buffering up as it means that the rate that you are pushing data out at is higher than you can actually write it over the wire. It's likely that you need to rate limit your outbound publishing of data. Increasing those buffers also increases the memory footprint of the application under load.

**Comment by mhchia** [comment]
@whyrusleeping Okay, I see. I will test further for this. 
Thank you a lot for the prompt reply 👍

**Comment by jamesray1** [comment]
Lots of questions are here: https://github.com/Drops-of-Diamond/diamond_drops/issues/34#issuecomment-388055539.

**Comment by whyrusleeping** [comment]
@jamesray1 great! I'll take a stab at answering them, though that post is a little hard to parse. Would it help for you and I to have a video call to talk through some things?

**Comment by whyrusleeping** [comment]
Just a heads up (cc @mhchia ) we merged a refactor to go-libp2p that improves and cleans up a lot of things, as well as makes adding new transport types (like QUIC) much easier. You might have to adjust some of your code slightly (the libp2p constructor in go-libp2p probably). @Stebalien will be doing a writeup of the changes very soon.

Sorry for any inconvenience.

**Comment by mhchia** [comment]
@whyrusleeping 
No problem! I will modify the code according to the change.
Thank you a lot for the reminder.

**Comment by Stebalien** [comment]
@mhchia https://github.com/libp2p/go-libp2p/pull/347

**Comment by mhchia** [comment]
sorry I have some questions again.
1. About the roadmap of gossipsub, I tried to find the information but the ones I found are relatively old or scattered ( https://github.com/libp2p/research-pubsub/issues/9#issue-171152811 ), is there a recent and future roadmap for gossipsub?
2. The [high-level spec](https://github.com/vyzo/gerbil-simsub) for current gossipsub https://github.com/libp2p/go-floodsub/pull/67 . In the spec it is metioned that 

> possible advanced gossipsub routers e.g. an epidemic broadcast tree can be used to optimize the overlay.

 I assume this is some similar approach mentioned in the [RFC](https://github.com/vyzo/research-pubsub/blob/a6cfdd3fec9d595a3f4cb56e94642e77b4dca13f/epidemic-pubsub.md) or particularly the paper epidemic broadcast trees. For example, IIUC from the code and the high-level spec, `GRAFT` and `PRUNE` are sent in heartbeat, but not sent when receiving `IHAVE` to heal/prune the mesh. Will this kind of optimization, or other features mentioned in the document be implemented in the future?

3. Is there benchmark for the https://github.com/libp2p/go-floodsub/pull/67 or https://github.com/libp2p/go-floodsub available?

4. Though it is asked in https://github.com/libp2p/go-floodsub/issues/77#issuecomment-387368885 before, I'm still excited about the testnet for gossipsub, and curious if it is in the roadmap.

Really appreciate for your help! Thanks a lot.

**Comment by whyrusleeping** [comment]
Hey @mhchia 

>  is there a recent and future roadmap for gossipsub?

We don't have anything written down right now. We are currently in the process of writing down all of our primary goals and figure out a roadmap across libp2p. A robust gossip layer on libp2p is a P0 priority for Filecoin, so that will be high on the list for us. (cc @mgoelzer re roadmapping)

> Will this kind of optimization, or other features mentioned in the document be implemented in the future?

Yes, those sorts of things will likely be implemented in the future. Our approach right now is to build out the pieces that are the common denominators of future advanced optimizations, so the gossipsub protocol with its GRAFTs PRUNEs IWANT and IHAVE messages allows these more interesting protocols to be built. Its worth noting though, that different approaches are better optimized for different scenarios, and just implementing one of them would hamper progress for other use cases. One idea would be that the particular routing implementation would be specified in the topic descriptor.

> Is there benchmark for the libp2p/go-floodsub#67 or https://github.com/libp2p/go-floodsub available?

No, not yes for gossipsub. For floodsub we have some synthetic benchmarks for maximum throughput, but those assume optimal conditions. Building out a testing framework for these protocols is pretty high priority for us.

> Though it is asked in libp2p/go-floodsub#77 (comment) before, I'm still excited about the testnet for gossipsub, and curious if it is in the roadmap.

Yeah! I'm hoping to get the gossipsub PR merged pretty soon. It's really just blocked on my and @Stebalien's time right now for final reviews and validation. Hoping to get to that next week.

**Comment by mhchia** [comment]
@whyrusleeping 
Thank you so much for the detailed reply!

> A robust gossip layer on libp2p is a P0 priority for Filecoin

Does that mean the whole gossipsub will be used as the gossip layer? or just part of it, or another implementation?

> Will this kind of optimization, or other features mentioned in the document be implemented in the future?
Yes, those sorts of things will likely be implemented in the future. Our approach right now is to build out the pieces that are the common denominators of future advanced optimizations, so the gossipsub protocol with its GRAFTs PRUNEs IWANT and IHAVE messages allows these more interesting protocols to be built. Its worth noting though, that different approaches are better optimized for different scenarios, and just implementing one of them would hamper progress for other use cases.

Thank you for sharing this information. It seems a lot to be considered and is complicated.

> One idea would be that the particular routing implementation would be specified in the topic descriptor.

Do you mean like specifying a message is routed by floodsub/gossipsub/… ? or is this “routing” means the peer routing?

Some questions related to the underlying network just came to my mind, and I'm sorry if they are dumb questions. Just want to be more knowledgeable of gossipsub.

1. Is there any requirement for the overlay network which the gossipsub is based on?
2. Can the peer sampling service under current/future gossipsub be either structured or unstructured network? (listed in Structured vs Unstructured networks, https://github.com/libp2p/research-pubsub/issues/9#issue-171152811 )
3. Does the term "random overlay network" mentioned in [epidemic broadcast trees](http://homepages.gsd.inesc-id.pt/~jleitao/pdf/srds07-leitao.pdf) mean “unstructured network”?
4. If we use DHT(structured network) to do peer routing, will there be problems for gossipsub?

Thanks again for all the answers. They really help a lot.

**Comment by whyrusleeping** [comment]
> Does that mean the whole gossipsub will be used as the gossip layer? or just part of it, or another implementation?

We're still investigating, and figuring out what our exact acceptance criteria here will be, but the idea is that the gossipsub protocol extensions (the IHAVE, IWANT, GRAFT, PRUNE) allow us to plug in any sort of underlying peer selection strategy. Currently, gossipsub is just the randomized limited peer selection, but its very easy to change.

> Do you mean like specifying a message is routed by floodsub/gossipsub/… ? or is this “routing” means the peer routing?

Yeah, like that, but in the future nodes might support multiple more advanced routing schemes, so you could select between floodsub, gossipsub, epidemic broadcast trees, or some other algorithm optimized for your usecase. (to be implemented)

> Is there any requirement for the overlay network which the gossipsub is based on?

Thats a rather broad question, but i think 'not really' is the answer. Right now we assume that messages are sent over a reliable channel, but i'm sure we could easily make things work in a lossy environment.

> Can the peer sampling service under current/future gossipsub be either structured or unstructured network? (listed in Structured vs Unstructured networks, libp2p/research-pubsub#9 (comment) )

You can definitely optimize things for a structured network, like a DHT, but at that point It wouldnt be gossipsub anymore, but some optimized version of it. The code here is designed to be flexible enough to allow you to do things just like that.

> If we use DHT(structured network) to do peer routing, will there be problems for gossipsub?

Depending on what you mean by peer routing, no. We use a DHT for peer routing, where peer routing means 'the ability to look up addresses for peers'.

**Comment by mhchia** [comment]
@whyrusleeping 
Thank you so much! It's really kind of you that solve my questions.

> No, not yes for gossipsub. For floodsub we have some synthetic benchmarks for maximum throughput, but those assume optimal conditions.

Sorry, I overlooked this answer before. Are the benchmarks available? It will be great if the maximum throughput is known.

> Depending on what you mean by peer routing, no. We use a DHT for peer routing, where peer routing means 'the ability to look up addresses for peers'.

Yep, I mean looking up the IP/port info for peers. My concern was that it seemed the papers only referred to the underlying network as "random" or "unstructured" overlay network, but not "structured" network. I used DHT for peer routing, so wanted to know if that will affect anything of the gossipsub which samples peers from it. Now my understanding is using structured overlay network as the underlying network is fine, and we can do further optimization based on its feature(like the XOR distance), but this kind of optimization will be specific to this network. Is this correct?

Big thanks and happy weekend:)

**Comment by whyrusleeping** [comment]
> Are the benchmarks available?

I thought they were just in the _test.go file in the repo, but it seems they arent there. All it was testing was essentially serialization stuff between simulated hosts, and we were able to get on the order of a 100k messages per second IIRC. It was more than enough to saturate your network pipes. We should really add some serious over-the-network type benchmarks for that all.

> we can do further optimization based on its feature(like the XOR distance), but this kind of optimization will be specific to this network. Is this correct?

Yep, that sounds right to me.

> Big thanks and happy weekend:)

You too :)

**Comment by jamesray1** [comment]
> I thought they were just in the _test.go file in the repo, but it seems they arent there. 

I remember seeing tests for the gerbil implementation.

**Comment by jamesray1** [comment]
Also FYI I answered some questions in this wiki [here](https://github.com/Drops-of-Diamond/diamond_drops/wiki/Design-with-libp2p).

**Comment by jamesray1** [comment]
For gossipsub, it will be interesting to see how well it would work for Ethereum without incentives to relay messages (which is critical for the protocol).

**Comment by whyrusleeping** [comment]
@jamesray1 thats a very interesting thought. How does this work in ethereums current gossip implementation?

**Comment by jamesray1** [comment]
There are no incentives for relaying in Ethereum. However, there is a concern that that is not sustainable, and that as the network scales it could lead to a tragedy of the commons due to an unincentivized resource/process.

**Comment by jamesray1** [comment]
Also I'm not an expert on the [heartbleed bug](http://heartbleed.com/) but implementations of a heartbeat (as in gossipsub) should be scrutinized to prevent leaks.

**Comment by whyrusleeping** [comment]
@jamesray1 heh, i'm not super worried about a buffer overread in go. Especially since our heartbeats are strictly a one-way function (broadly, heartbleed was caused by one side being able to specify a length of data that was not validated against how much data was actually received)

**Comment by whyrusleeping** [comment]
That said, we are working on organizing security audits for all the core components of libp2p.
